src/modules/rlm_smtp/rlm_smtp.c

```diff
@@ -0,0 +1,1107 @@
+	{ .out = &attr_smtp_header, .name = "SMTP-Mail-Header", .type = FR_TYPE_STRING, .dict = &dict_freeradius },
+	{ .out = &attr_smtp_body, .name = "SMTP-Mail-Body", .type = FR_TYPE_STRING, .dict = &dict_freeradius },
+	{ .out = &attr_smtp_attachment_file, .name = "SMTP-Attachments", .type = FR_TYPE_STRING, .dict = &dict_freeradius },
+	{ NULL },
+};
+
+typedef struct {
+	char const		*uri;			//!< URI of smtp server
+	char const		*template_dir;		//!< The directory that contains all email attachments
+	char const		*sender_address;	//!< The address used to generate the FROM: header
+	char const		*envelope_address;	//!< The address used to send the message
+	tmpl_t 			**attachments;		//!< The attachmen ts to be set
+	tmpl_t 			**recipient_addrs;	//!< Comma separated list of emails. Overrides elements in to, cc, bcc
+	tmpl_t 			**to_addrs;		//!< Comma separated list of emails to be listed in TO:
+	tmpl_t 			**cc_addrs;		//!< Comma separated list of emails to be listed in CC:
+	tmpl_t 			**bcc_addrs;		//!< Comma separated list of emails not to be listed
+	fr_time_delta_t 	timeout;		//!< Timeout for connection and server response
+	fr_curl_tls_t		tls;			//!< Used for handled all tls specific curl components
+	char const		*name;			//!< Auth-Type value for this module instance.
+	fr_dict_enum_t		*auth_type;
+	vp_map_t		*header_maps;			//!< Attribute map used to process header elements
+	CONF_SECTION		*cs;
+	bool 			set_date;
+} rlm_smtp_t;
+
+typedef struct {
+	rlm_smtp_t const    	*inst;		//!< Instance of rlm_smtp.
+	fr_curl_handle_t    	*mhandle;	//!< Thread specific multi handle.  Serves as the dispatch and coralling structure for smtp requests
+} rlm_smtp_thread_t;
+
+/*
+ *	Holds the context for parsing the email elements
+ */
+typedef struct {
+	REQUEST			*request;
+	fr_curl_io_request_t	*randle;
+	fr_cursor_t		cursor;
+	fr_cursor_t		body_cursor;
+	fr_dbuff_t		vp_in;
+	struct curl_slist	*recipients;
+	struct curl_slist	*header;
+	struct curl_slist 	*body_header;
+	fr_time_t 		time;
+	char 			time_str[60];
+	curl_mime		*mime;
+} fr_mail_ctx;
+
+/*
+ *	A mapping of configuration file names to internal variables.
+ */
+static const CONF_PARSER module_config[] = {
+	{ FR_CONF_OFFSET("uri", FR_TYPE_STRING, rlm_smtp_t, uri) },
+	{ FR_CONF_OFFSET("template_directory", FR_TYPE_STRING, rlm_smtp_t, template_dir) },
+	{ FR_CONF_OFFSET("attachments", FR_TYPE_TMPL | FR_TYPE_MULTI, rlm_smtp_t, attachments), .dflt = "&SMTP-Attachments", .quote = T_BARE_WORD},
+	{ FR_CONF_OFFSET("sender_address", FR_TYPE_STRING, rlm_smtp_t, sender_address) },
+	{ FR_CONF_OFFSET("envelope_address", FR_TYPE_STRING, rlm_smtp_t, envelope_address) },
+	{ FR_CONF_OFFSET("recipients", FR_TYPE_TMPL | FR_TYPE_MULTI, rlm_smtp_t, recipient_addrs), .dflt = "&SMTP-Recipients", .quote = T_BARE_WORD},
+	{ FR_CONF_OFFSET("TO", FR_TYPE_TMPL | FR_TYPE_MULTI, rlm_smtp_t, to_addrs), .dflt = "&SMTP-TO", .quote = T_BARE_WORD},
```

Comment:
Same for the rest of these.  Should be looking at all instances of the attribute, not just the first.